modules/RTC/RTCUtils.js

```diff
@@ -1287,167 +1287,142 @@ class RTCUtils extends Listenable {
      * @returns {*} Promise object that will receive the new JitsiTracks
      */
     obtainAudioAndVideoPermissions(options = {}) {
-        const self = this;
-
         const dsOptions = {
             ...options.desktopSharingExtensionExternalInstallation,
             desktopSharingSources: options.desktopSharingSources
         };
 
-        return new Promise((resolve, reject) => {
-            const successCallback = function(stream) {
-                resolve(handleLocalStream(stream, options.resolution));
-            };
+        options.devices = options.devices || [ 'audio', 'video' ];
+        options.resolution = options.resolution || '720';
 
-            options.devices = options.devices || [ 'audio', 'video' ];
-            options.resolution = options.resolution || '720';
+        const requestingDesktop = options.devices.includes('desktop');
 
-            if (!screenObtainer.isSupported()
-                && options.devices.indexOf('desktop') !== -1) {
-                reject(new Error('Desktop sharing is not supported!'));
-            }
-            if (browser.isFirefox()
-
-                    // XXX The react-native-webrtc implementation that we
-                    // utilize on React Native at the time of this writing does
-                    // not support the MediaStream constructors defined by
-                    // https://www.w3.org/TR/mediacapture-streams/#constructors
-                    // and instead has a single constructor which expects (an
-                    // NSNumber as) a MediaStream ID.
-                    || browser.isReactNative()
-                    || browser.isTemasysPluginUsed()) {
-                const GUM = function(device, s, e) {
-                    this.getUserMediaWithConstraints(device, s, e, options);
-                };
+        if (requestingDesktop && !screenObtainer.isSupported()) {
+            return Promise.reject(
+                new Error('Desktop sharing is not supported!'));
+        }
+
+        let gumPromise;
+
+        if (browser.supportsMediaStreamConstructor()) {
+            options.devices = options.devices.filter(device =>
+                device !== 'desktop');
+
+            const promiseWrappedGum = (requestedDevices, gumOptions) =>
+                new Promise((resolve, reject) => {
+
+                    if (requestedDevices.length) {
+                        this.getUserMediaWithConstraints(
+                            requestedDevices,
+                            resolve,
+                            reject,
+                            gumOptions);
+                    } else {
+                        resolve(null);
+                    }
+                });
+
+            gumPromise = promiseWrappedGum(options.devices, options)
+                .then(avStream => {
+                    const missingTracks
+                        = this._getMissingTracks(options.devices, avStream);
```

Comment:
do we know that the scenario that required this error handling still exists?  i wonder if it's something we could get rid of.

Reply:
The scenario is that we're trying to obtain audio and video stream with one GUM request, but only video will fail, because user granted mic, but not cam (and vice versa). So I'd say it's quite common and still needs to be handled.

Reply:
ok, i wasn't thinking so much that the case didn't exist, but that the need for our special error-handling (as opposed to the built in error) was no longer needed, but maybe that isn't the case.